NMS repackaging

Proximyst opened the pull request:
This PR blocks master.

Big thanks to @jpenilla for his work on the initial update.

jpenilla committed:
Update scripts for NMS repackaging

jpenilla committed:
Updated Upstream (CraftBukkit/Spigot)

Doesn't compile yet.

CraftBukkit Changes:
90d6905b Repackage NMS
69cf961d Repackage patches

Spigot Changes:
79d53c28 Repackage NMS

Proximyst committed:
More work

Proximyst committed:
It compiles

Proximyst committed:
more work

Proximyst committed:
Change cache keys

Proximyst committed:
Welcome to 1.16.5-R0.2

Proximyst committed:
Use revision file

Proximyst committed:
Apply fixups

Proximyst committed:
It's mojang math, not minecraft math

Proximyst committed:
Fix class preloading

Proximyst committed:
Actually fix preloading this time

Proximyst committed:
Rename work dir

Proximyst committed:
Link correctly

MiniDigger reviewed (approved):
seems fine. if its broken, we blame md5

Proximyst committed:
Fix macOS/BSD support

Proximyst committed:
Don't remove the .java

Proximyst committed:
Replace ** with * (BSD/macOS)

Proximyst committed:
script & POM fix

Proximyst committed:
Revert de5f4e469...c270abe96

jpenilla commented on scripts/decompile.sh:
```diff
@@ -73,7 +73,7 @@ if [ ! -d "$classdir" ]; then
     cd "$classdir"
     set +e
-    jar xf "$decompiledir/$minecraftversion-mapped.jar" net/minecraft/server
+    jar xf "$decompiledir/$minecraftversion-mapped.jar" net/minecraft
```
```suggestion
    jar xf "$decompiledir/$minecraftversion-mapped.jar" net/minecraft com/mojang/math
```

jpenilla reviewed (changes requested):
We also need to extract the classes in `com.mojang.math`. importmcdev probably needs changes as well to work with `com.mojang.math` classes when/if we want to patch them.

Proximyst committed:
rebase

Proximyst committed:
fix compile

jpenilla commented on Spigot-Server-Patches/0534-Fix-SPIGOT-5989.patch:
```diff
@@ -0,0 +1,5849 @@
+new file mode 100644
+index 0000000000000000000000000000000000000000..12d6e26ec0b25bef42004c6632fca404c8b0c30b
+--- /dev/null
++++ b/t
```
t

Proximyst committed:
wtf is this t file -jmp

Proximyst committed:
Updated Upstream (Bukkit/CraftBukkit)

Upstream has released updates that appear to apply and compile correctly.
This update has not been tested by PaperMC and as with ANY update, please do your own testing

Bukkit Changes:
4e5f46cb Update ASM version

CraftBukkit Changes:
d54b500d Prepare for Java 16

zachbr reviewed (approved):
What a fun diff. Reviews a little better in git directly but not much.

I have confirmed this works nicely in Ubuntu 20.04 and Windows 10 in git bash after clearing the work dir (`/paper clean`). Jar layout looks as I'd expect and plugins look to be working correctly as expected.

Thanks all. Please ping a core team member immediately before merge so we can make sure the CI gets its workspace cleaned once.
